internal/netsync/manager.go

```diff
@@ -1335,167 +1303,126 @@ func (m *SyncManager) handleInvMsg(imsg *invMsg) {
-						"block: %v", err)
-					continue
-				}
-				locator := chainBlockLocatorToHashes(blkLocator)
-				err = peer.PushGetBlocksMsg(locator, orphanRoot)
-				if err != nil {
-					log.Errorf("Failed to push getblocksmsg for orphan chain: "+
-						"%v", err)
-				}
+			// Ignore block announcements for blocks that are from peers other
+			// than the sync peer before the chain is current or are otherwise
+			// not needed, such as when they're already available.
+			//
+			// This helps prevent fetching a mass of orphans.
+			if (!fromSyncPeer && !isCurrent) || !m.needBlock(&iv.Hash) {
 				continue
 			}
 
-			// We already have the final block advertised by this
-			// inventory message, so force a request for more.  This
-			// should only happen if we're on a really long side
-			// chain.
-			if i == lastBlock {
-				// Request blocks after this one up to the
-				// final one the remote peer knows about (zero
-				// stop hash).
-				blkLocator := m.cfg.Chain.BlockLocatorFromHash(&iv.Hash)
-				locator := chainBlockLocatorToHashes(blkLocator)
-				err = peer.PushGetBlocksMsg(locator, &zeroHash)
-				if err != nil {
-					log.Errorf("PEER: Failed to push getblocksmsg: %v", err)
-				}
-			}
-		}
-	}
-
-	// Request as much as possible at once.
-	numRequested := 0
-	gdmsg := wire.NewMsgGetData()
-	for _, iv := range requestQueue {
-		switch iv.Type {
-		case wire.InvTypeBlock:
-			// Request the block if there is not already a pending
-			// request.
+			// Request the block if there is not one already pending.
 			if _, exists := m.requestedBlocks[iv.Hash]; !exists {
 				limitAdd(m.requestedBlocks, iv.Hash, maxRequestedBlocks)
 				limitAdd(peer.requestedBlocks, iv.Hash, maxRequestedBlocks)
-				gdmsg.AddInvVect(iv)
-				numRequested++
+				requestQueue = append(requestQueue, iv)
 			}
 
 		case wire.InvTypeTx:
-			// Request the transaction if there is not already a
-			// pending request.
+			// Add the tx to the cache of known inventory for the peer.  This
+			// helps avoid sending transactions to the peer that it is already
+			// known to have.
+			peer.AddKnownInventory(iv)
+
+			// Ignore transaction announcements before the chain is current or
+			// are otherwise not needed, such as when they were recently
+			// rejected or are already known.
+			//
+			// Transaction announcements are based on the state of the fully
+			// synced ledger, so they are likely to be invalid before the chain
+			// is current.
+			if !isCurrent || !m.needTx(&iv.Hash) {
+				continue
+			}
+
+			// Request the transaction if there is not one already pending.
 			if _, exists := m.requestedTxns[iv.Hash]; !exists {
 				limitAdd(m.requestedTxns, iv.Hash, maxRequestedTxns)
 				limitAdd(peer.requestedTxns, iv.Hash, maxRequestedTxns)
-				gdmsg.AddInvVect(iv)
-				numRequested++
+				requestQueue = append(requestQueue, iv)
 			}
 		}
+	}
 
+	if lastBlock != nil {
+		// The block is an orphan block that we already have.  When the
```

Comment:
Since this comment relates to the check below I think it should be reworded as it is true of if `isKnownOrphan` is true. Perhaps instead of `The block is an orphan block that we already have ...` it could be `If the block is an orphan block that we already have...`

Reply:
Could also keep it as is but move the comments into the if block checking `isKnownOrphan`.